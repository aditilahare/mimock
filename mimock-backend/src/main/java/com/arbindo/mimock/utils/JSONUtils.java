package com.arbindo.mimock.utils;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import lombok.extern.log4j.Log4j2;
import org.apache.logging.log4j.Level;

import java.util.HashMap;
import java.util.Map;
import java.util.Objects;

@SuppressWarnings("unchecked")
@Log4j2
public class JSONUtils {

    public static Map<String, Object> convertJSONStringToMap(String jsonString) {
        ObjectMapper objectMapper = new ObjectMapper();

        if (jsonString == null || jsonString.isBlank()) {
            log.log(Level.ERROR, "JSON string is blank");
            return null;
        }

        try {
            log.log(Level.INFO, "Generating map from JSON string");
            return objectMapper.readValue(jsonString, Map.class);
        } catch (JsonProcessingException e) {
            log.log(Level.ERROR, "Error occurred while parsing JSON string : " + e.getMessage());
            return null;
        }
    }

    public static String convertMapToJSONString(Map<String, Object> jsonMap) {
        ObjectMapper objectMapper = new ObjectMapper();

        if (jsonMap == null || jsonMap.isEmpty()) {
            log.log(Level.ERROR, "JSON map is empty");
            return null;
        }

        try {
            log.log(Level.INFO, "Generating json string from map");
            return objectMapper.writeValueAsString(jsonMap);
        } catch (JsonProcessingException e) {
            log.log(Level.ERROR, "Error occurred while converting json map to string : " + e.getMessage());
            return null;
        }
    }

    public static Map<String, Object> convertJSONStringToMapWithLowerCaseKeys(String jsonString) {
        if (jsonString == null || jsonString.isBlank()) {
            log.log(Level.ERROR, "JSON string is blank");
            return null;
        }

        return changeKeyCaseToLower(Objects.requireNonNull(convertJSONStringToMap(jsonString)));
    }

    private static Map<String, Object> changeKeyCaseToLower(Map<String, Object> jsonMap) {
        log.log(Level.INFO, "Converting map keys to lower case");
        Map<String, Object> mapWithChangedCase = new HashMap<>();
        for (Map.Entry<String, Object> header : jsonMap.entrySet()) {
            mapWithChangedCase.put(header.getKey().toLowerCase(), header.getValue());
        }

        return mapWithChangedCase;
    }
}
