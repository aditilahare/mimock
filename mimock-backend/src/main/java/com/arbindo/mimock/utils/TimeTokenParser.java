package com.arbindo.mimock.utils;

import lombok.extern.log4j.Log4j2;
import org.apache.logging.log4j.Level;

@Log4j2
public class TimeTokenParser {
    private static final long DEFAULT_DURATION = 3600L;

    private enum TimeMetrics {
        YEAR('y', 365 * 24),
        MONTH('M', 30 * 24),
        WEEK('w', 7 * 24),
        DAY('d', 24),
        HOUR('h', 1),
        MINUTE('m', 1.0 / 60.0),
        SECOND('s', 1.0 / 3600.0);

        char token;
        double hourValue;

        TimeMetrics(char token, double hourValue) {
            this.token = token;
            this.hourValue = hourValue;
        }
    }

    public static long durationInSeconds(String timeToken) {
        try {
            char token = timeToken.charAt(timeToken.length() - 1);
            long convertedDuration = Long.parseLong(getDurationFrom(timeToken));

            for (TimeMetrics timeMetric : TimeMetrics.values()) {
                if (timeMetric.token == token) {
                    return (long) (convertedDuration * timeMetric.hourValue * 3600);
                }
            }
        } catch (Exception e) {
            log.log(Level.ERROR, e.getMessage());
        }

        return DEFAULT_DURATION;
    }

    private static String getDurationFrom(String timeToken) {
        return timeToken.substring(0, timeToken.length() - 1);
    }
}
