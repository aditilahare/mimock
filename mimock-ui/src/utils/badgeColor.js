export const decideBadgeColor = (httpMethod) => {
	const colorMap = {
		GET: 'blue-500',
		POST: 'green-500',
		PUT: 'yellow-500',
		DELETE: 'red-500',
		PATCH: 'yellow-900',
		CONNECT: 'pink-500',
		HEAD: 'indigo-500',
		TRACE: 'violet-500',
		OPTIONS: 'gray-500',
	};

	const color = colorMap[httpMethod] || 'gray-500';

	return `border-2 border-${color} text-${color}`;
};
