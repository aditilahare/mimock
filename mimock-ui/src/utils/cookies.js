import { Cookies } from 'react-cookie';

const cookies = new Cookies();

export const setCookiesWithExpiry = async (name, value, expiryInSeconds) => {
	document.cookie = `${name}=${value}; expires=${new Date(
		Date.now() + parseInt(expiryInSeconds) * 1000
	)}; path=/`;

	return Promise.resolve(cookies.get(name));
};
