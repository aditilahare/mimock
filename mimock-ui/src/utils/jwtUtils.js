import { Buffer } from 'buffer';
import { Cookies } from 'react-cookie';
import { globalConstants } from 'src/constants/globalConstants';

export const getUserDetails = () => {
	const authToken = new Cookies().get(globalConstants.AUTH_TOKEN_COOKIE_NAME);

	if (!authToken) {
		throw new Error('Auth token does not exist');
	}

	const splitToken = authToken.toString().split('.');
	if (splitToken?.length < 3) {
		throw new Error('Auth token is invalid');
	}

	const userDetails = Buffer.from(authToken.split('.')[1], 'base64').toString(
		'ascii'
	);

	return JSON.parse(userDetails);
};
