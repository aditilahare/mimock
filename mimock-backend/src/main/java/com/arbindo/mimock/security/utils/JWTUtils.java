package com.arbindo.mimock.security.utils;

import com.arbindo.mimock.utils.TimeTokenParser;
import io.jsonwebtoken.Claims;
import io.jsonwebtoken.Jwts;
import lombok.Getter;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.stereotype.Service;

import javax.crypto.spec.SecretKeySpec;
import java.nio.charset.StandardCharsets;
import java.util.Date;
import java.util.HashMap;
import java.util.function.Function;

@Service
public class JWTUtils {

    @Getter
    @Value("${app.security.jwt-expiry-duration}")
    private String jwtExpiryDuration;

    @Value("${app.security.jwt-secret-key}")
    private String jwtSecretKey;

    public String extractUsername(String token) {
        return extractClaim(token, Claims::getSubject);
    }

    public Date extractExpiration(String token) {
        return extractClaim(token, Claims::getExpiration);
    }

    public <T> T extractClaim(String token, Function<Claims, T> claimsResolver) {
        final Claims claims = extractAllClaims(token);
        return claimsResolver.apply(claims);
    }

    private Claims extractAllClaims(String token) {
        return Jwts.parserBuilder().setSigningKey(getSigningKey()).build().parseClaimsJws(token).getBody();
    }

    public Boolean isTokenExpired(String token) {
        return extractExpiration(token).before(new Date());
    }

    public Date getTokenExpiryTimestamp() {
        return new Date(System.currentTimeMillis() + (TimeTokenParser.durationInSeconds(jwtExpiryDuration) * 1000));
    }

    private SecretKeySpec getSigningKey() {
        return new SecretKeySpec(jwtSecretKey.getBytes(StandardCharsets.UTF_8), "HmacSHA256");
    }

    public String generateJWT(UserDetails userDetails, HashMap<String, Object> claims) {
        return Jwts.builder()
                .setClaims(claims)
                .setSubject(userDetails.getUsername())
                .setHeaderParam("iat", new Date().toString())
                .setExpiration(getTokenExpiryTimestamp())
                .signWith(getSigningKey())
                .compact();
    }

    public Boolean validateToken(String token, UserDetails userDetails) {
        final String username = extractUsername(token);
        return username.equals(userDetails.getUsername());
    }
}
