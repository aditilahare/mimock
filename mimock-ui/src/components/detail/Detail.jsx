import React, { useState, useEffect, useRef } from 'react';
import { useLocation } from 'react-router-dom';
import { getMockById } from 'services/mockManagement/mockManagement.service';
import { Cookies } from 'react-cookie';
import { globalConstants } from 'constants/globalConstants';
import { DetailContainer } from './Detail.style';
import DetailToolbar from './detailtoolbar/DetailToolbar';
import DetailHeader from './detailheader';
import DetailContentViewer from './detailcontentviewer';
import { decideBadgeColor } from 'utils/badgeColor.js';

function Detail() {
	const [mock, setMock] = useState(null);
	const [badgeColor, setBadgeColor] = useState('');

	const location = useLocation();
	const cookies = new Cookies();
	const authCookieRef = useRef('');
	const csrfCookieRef = useRef('');

	useEffect(() => {
		authCookieRef.current = cookies.get(globalConstants.AUTH_TOKEN_COOKIE_NAME);
		csrfCookieRef.current = cookies.get(globalConstants.XSRF_COOKIE_NAME);
	}, []);

	useEffect(() => {
		if (location.pathname != undefined) {
			const mockId = location.pathname.split('/detail/')[1];
			getMockById(mockId, authCookieRef)
				.then((res) => {
					console.log(res.data);
					setMock(res.data.data);
					const color = decideBadgeColor(res.data.data.httpMethod.method);
					setBadgeColor(color);
				})
				.catch((err) => console.log(err));
		}
	}, []);

	return (
		<DetailContainer data-testid='detail-section'>
			<DetailToolbar />
			<If condition={mock != null}>
				<DetailHeader mock={mock} badgeColor={badgeColor} />
				<DetailContentViewer mock={mock} />
			</If>
		</DetailContainer>
	);
}

export default Detail;
