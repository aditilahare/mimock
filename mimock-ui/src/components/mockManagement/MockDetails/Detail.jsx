import React, { useState, useEffect, useRef } from 'react';
import { useLocation, useNavigate } from 'react-router-dom';
import {
	getMockById,
	archiveMock,
	unarchiveMock,
	deleteMockById,
} from 'services/mockManagement/mockManagement.service';
import { Cookies } from 'react-cookie';
import { globalConstants } from 'constants/globalConstants';
import { DetailContainer } from './Detail.style';
import DetailToolbar from './detailtoolbar/DetailToolbar';
import DetailHeader from './detailheader';
import DetailContentViewer from './detailcontentviewer';
import { decideBadgeColor } from 'utils/badgeColor.js';
import { notificationTypes } from 'constants/notificationConstants';
import useNotification from 'hooks/useNotification';
import { ConfirmationModal } from 'components/common/Modals';

function Detail() {
	const [mock, setMock] = useState(null);
	const [mockId, setMockId] = useState(null);
	const [badgeColor, setBadgeColor] = useState('');
	const [showDeletionModal, setShowDeletionModal] = useState(false);
	const [deletingMock, setDeletingMock] = useState(false);

	const navigate = useNavigate();

	const location = useLocation();
	const cookies = new Cookies();
	const authCookieRef = useRef('');
	const csrfCookieRef = useRef('');

	useEffect(() => {
		authCookieRef.current = cookies.get(globalConstants.AUTH_TOKEN_COOKIE_NAME);
		csrfCookieRef.current = cookies.get(globalConstants.XSRF_COOKIE_NAME);
	}, []);

	useEffect(() => {
		if (location.pathname != undefined) {
			const mockIdString = location.pathname.split('/detail/')[1];
			setMockId(mockIdString);
			getMockById(mockIdString, authCookieRef)
				.then((res) => {
					setMock(res.data.data);
					const color = decideBadgeColor(res.data.data.httpMethod.method);
					setBadgeColor(color);
				})
				.catch((err) => console.log(err));
		}
	}, []);

	const performArchiveMockOperation = () => {
		archiveMock(mockId, authCookieRef)
			.then((res) => {
				if (res.status == 200) {
					navigate(`/mimock-ui/mocks`);
					useNotification({
						type: notificationTypes.NOTIFICATION_TYPE_SUCCESS,
						title: 'Archive success',
						message: `${res.data.data.mockName}: ${res.data.message}`,
						animationIn: 'animate__bounceIn',
						animationOut: 'animate__bounceOut',
					});
				}
				console.log(res.status, res.data);
			})
			.catch((err) => {
				console.log(err);
				useNotification({
					type: notificationTypes.NOTIFICATION_TYPE_ERROR,
					title: 'Failed to archive mock',
					message: `${err.message}`,
					animationIn: 'animate__bounceIn',
					animationOut: 'animate__bounceOut',
				});
			});
	};

	const performUnarchiveMockOperation = () => {
		unarchiveMock(mockId, authCookieRef)
			.then((res) => {
				if (res.status == 200) {
					navigate(`/mimock-ui/mocks`);
					useNotification({
						type: notificationTypes.NOTIFICATION_TYPE_SUCCESS,
						title: 'Unarchive success',
						message: `${res.data.data.mockName}: ${res.data.message}`,
						animationIn: 'animate__bounceIn',
						animationOut: 'animate__bounceOut',
					});
				}
				console.log(res.status, res.data);
			})
			.catch((err) => {
				console.log(err);
				useNotification({
					type: notificationTypes.NOTIFICATION_TYPE_ERROR,
					title: 'Failed to unarchive mock',
					message: `${err.message}`,
					animationIn: 'animate__bounceIn',
					animationOut: 'animate__bounceOut',
				});
			});
	};

	const performDeleteMockOperation = () => {
		setShowDeletionModal(true);
	};

	const handleDeleteMock = async () => {
		setDeletingMock(true);
		await deleteMockById(mockId, authCookieRef)
			.then((res) => {
				setDeletingMock(false);
				setShowDeletionModal(false);
				if (res.status == 204) {
					navigate(`/mimock-ui/mocks`);
					useNotification({
						type: notificationTypes.NOTIFICATION_TYPE_SUCCESS,
						title: 'Deletion success',
						message: `Deleted resource successfully!`,
						animationIn: 'animate__bounceIn',
						animationOut: 'animate__bounceOut',
					});
				}
				console.log(res.status, res.data);
			})
			.catch((err) => {
				console.log(err);
				setDeletingMock(false);
				setShowDeletionModal(false);
				useNotification({
					type: notificationTypes.NOTIFICATION_TYPE_ERROR,
					title: 'Failed to delete mock',
					message: `${err.message}`,
					animationIn: 'animate__bounceIn',
					animationOut: 'animate__bounceOut',
				});
			});
	};

	const performEditMockOperation = () => {
		console.log('Edit Btn Clicked');
	};

	return (
		<DetailContainer data-testid='detail-section'>
			<If condition={mock != null}>
				<If condition={showDeletionModal}>
					<ConfirmationModal
						message={`Are you sure you want to delete mock "${mock.mockName}" ?`}
						cancelButtonLabel='Cancel'
						confirmButtonLabel='Delete Mock'
						loading={deletingMock}
						loadingMessage={`Deleting mock. Please wait...`}
						onConfirm={async () => {
							await handleDeleteMock();
						}}
						onCancel={() => {
							setShowDeletionModal(false);
						}}
					/>
				</If>
				<DetailToolbar
					status={mock.entityStatus.status}
					performArchiveMockOperation={performArchiveMockOperation}
					performUnarchiveMockOperation={performUnarchiveMockOperation}
					performDeleteMockOperation={performDeleteMockOperation}
					performEditMockOperation={performEditMockOperation}
				/>
				<DetailHeader mock={mock} badgeColor={badgeColor} />
				<DetailContentViewer mock={mock} />
			</If>
		</DetailContainer>
	);
}

export default Detail;
