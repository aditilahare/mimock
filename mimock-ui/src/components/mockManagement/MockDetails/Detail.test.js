import React from 'react';
import { render, act, fireEvent } from '@testing-library/react';
import { useLocation, useNavigate } from 'react-router-dom';
import {
	getMockById,
	archiveMock,
	unarchiveMock,
	deleteMockById,
} from 'services/mockManagement/mockManagement.service';
import { getUserDetails } from 'utils/jwtUtils';
import Detail from './Detail';

jest.mock('react-router-dom');
jest.mock('utils/jwtUtils');
jest.mock('services/mockManagement/mockManagement.service');

const actualNotification = jest.requireActual('hooks/useNotification');
jest.mock('react-notifications-component', () => {
	const Store = {
		addNotification: jest.fn(),
	};
	return { Store };
});

jest.useFakeTimers();

const data = {
	id: '4708bf6f-8122-4ca5-915f-1b098aa362b8',
	mockName: 'Test mock',
	description: 'Test mock',
	route: '/test1',
	httpMethod: {
		method: 'GET',
	},
	statusCode: 200,
	responseContentType: {
		contentType: 'application/json',
		description: 'JSON format',
		responseType: {
			name: 'TEXTUAL_RESPONSE',
		},
	},
	queryParams: 'q=ship',
	queryParamValues: {
		q: 'ship',
	},
	requestBodiesForMock: {
		id: 1,
		requestBody: {
			key: 'value',
		},
		createdAt: '2022-08-03T23:23:50.535995+05:30',
		updatedAt: '2022-08-03T23:23:50.535995+05:30',
		requestBodyType: {
			requestBodyType: 'application/json',
		},
	},
	createdAt: '2022-08-03T22:31:32.911138+05:30',
	updatedAt: '2022-08-07T21:19:41.046908+05:30',
	deletedAt: null,
	entityStatus: {
		status: 'NONE',
	},
	createdBy: 'Mimock Admin',
	modifiedBy: 'Mimock Admin',
	deleted: false,
	archived: false,
};

describe('Detail', () => {
	beforeEach(() => {
		useLocation.mockImplementation(() => ({
			pathname: '/detail/4708bf6f-8122-4ca5-915f-1b098aa362b8',
		}));

		getMockById.mockResolvedValue({
			data: {
				data,
			},
		});

		archiveMock.mockResolvedValue({
			data: {
				message: 'Mock archived successfully',
				data: {
					...data,
					entityStatus: {
						status: 'ARCHIVED',
					},
				},
			},
		});

		unarchiveMock.mockResolvedValue({
			data: {
				message: 'Mock unarchived successfully',
				data: {
					...data,
					entityStatus: {
						status: 'NONE',
					},
				},
			},
		});

		getUserDetails.mockImplementation(() => ({
			userRole: 'ROLE_ADMIN',
		}));
	});

	it('should render detail component for ADMIN', async () => {
		jest.spyOn(global, 'setTimeout');

		let tree;
		await act(async () => {
			tree = await render(<Detail />);
		});

		const { container, getByTestId, queryByTestId } = tree;

		expect(getByTestId('detail-container')).toBeInTheDocument();
		expect(getByTestId('detail-toolbar-container')).toBeInTheDocument();
		expect(queryByTestId('fetch-mock-details-error')).not.toBeInTheDocument();

		await act(async () => {
			await fireEvent.click(getByTestId('copy-url-button'));
		});

		await act(() => {
			jest.advanceTimersByTime(1500);
			expect(setTimeout).toHaveBeenCalledTimes(1);
			expect(setTimeout).toHaveBeenLastCalledWith(expect.any(Function), 1000);
		});
		expect(container).toMatchSnapshot();
	});

	it('should render detail component for VIEWER', async () => {
		getMockById.mockResolvedValue({
			data: {
				data: {
					...data,
					queryParams: '',
				},
			},
		});

		getUserDetails.mockImplementation(() => ({
			userRole: 'ROLE_VIEWER',
		}));

		let tree;
		await act(async () => {
			tree = await render(<Detail />);
		});

		const { container, getByTestId, queryByTestId } = tree;

		expect(getByTestId('detail-container')).toBeInTheDocument();
		expect(queryByTestId('operations-container')).not.toBeInTheDocument();
		expect(queryByTestId('fetch-mock-details-error')).not.toBeInTheDocument();

		expect(container).toMatchSnapshot();
	});

	it('should render ARCHIVED mock details', async () => {
		getMockById.mockResolvedValue({
			data: {
				data: {
					...data,
					entityStatus: {
						status: 'ARCHIVED',
					},
					archived: true,
				},
			},
		});

		let tree;
		await act(async () => {
			tree = await render(<Detail />);
		});

		const { container, getByTestId, queryByTestId } = tree;

		expect(getByTestId('detail-container')).toBeInTheDocument();
		expect(getByTestId('detail-toolbar-container')).toBeInTheDocument();
		expect(queryByTestId('fetch-mock-details-error')).not.toBeInTheDocument();

		expect(container).toMatchSnapshot();
	});

	it('should render DELETED mock details', async () => {
		getMockById.mockResolvedValue({
			data: {
				data: {
					...data,
					entityStatus: {
						status: 'DELETED',
					},
					deleted: true,
				},
			},
		});

		let tree;
		await act(async () => {
			tree = await render(<Detail />);
		});

		const { container, getByTestId, queryByTestId } = tree;

		expect(getByTestId('detail-container')).toBeInTheDocument();
		expect(getByTestId('warning-message-operations')).toBeInTheDocument();

		expect(queryByTestId('fetch-mock-details-error')).not.toBeInTheDocument();

		expect(container).toMatchSnapshot();
	});

	it('should display error banner when getMockById call fails', async () => {
		getMockById.mockRejectedValue(new Error('Error fetching mock details'));

		let tree;
		await act(async () => {
			tree = await render(<Detail />);
		});

		const { container, getByTestId, queryByTestId } = tree;

		expect(getByTestId('detail-container')).toBeInTheDocument();
		expect(getByTestId('fetch-mock-details-error')).toBeInTheDocument();

		expect(queryByTestId('detail-toolbar-container')).not.toBeInTheDocument();

		expect(container).toMatchSnapshot();
	});

	it('should display error when path is invalid', async () => {
		useLocation.mockImplementation(() => ({
			pathname: '/detail',
		}));

		getMockById.mockRejectedValue(new Error('Error fetching mock details'));

		let tree;
		await act(async () => {
			tree = await render(<Detail />);
		});

		const { container, getByTestId, queryByTestId } = tree;

		expect(getByTestId('detail-container')).toBeInTheDocument();
		expect(getByTestId('fetch-mock-details-error')).toBeInTheDocument();

		expect(queryByTestId('detail-toolbar-container')).not.toBeInTheDocument();

		expect(container).toMatchSnapshot();
	});

	it('should hide toolbar when getting user details fail', async () => {
		getUserDetails.mockImplementation(() => {
			throw new Error('Error fetching user details');
		});

		let tree;
		await act(async () => {
			tree = await render(<Detail />);
		});

		const { container, getByTestId, queryByTestId } = tree;

		expect(getByTestId('detail-container')).toBeInTheDocument();

		expect(queryByTestId('fetch-mock-details-error')).not.toBeInTheDocument();
		expect(queryByTestId('operations-container')).not.toBeInTheDocument();

		expect(container).toMatchSnapshot();
	});

	it('should archive mock', async () => {
		const notificationSpy = jest.spyOn(actualNotification, 'default');

		let tree;
		await act(async () => {
			tree = await render(<Detail />);
		});

		const { getByTestId } = tree;

		const archiveButton = getByTestId('archive-btn');
		await act(async () => {
			await fireEvent.click(archiveButton);
		});

		expect(archiveMock).toHaveBeenCalledTimes(1);
		expect(notificationSpy).toHaveBeenCalledTimes(1);
		expect(notificationSpy).toHaveBeenCalledWith({
			animationIn: 'animate__bounceIn',
			animationOut: 'animate__bounceOut',
			message: 'Test mock: Mock archived successfully',
			title: 'Archive success',
			type: 'success',
		});
	});

	it('should show error notification when archiving mock fails', async () => {
		const notificationSpy = jest.spyOn(actualNotification, 'default');
		archiveMock.mockRejectedValue(new Error('Failed to archive mock'));

		let tree;
		await act(async () => {
			tree = await render(<Detail />);
		});

		const { getByTestId } = tree;

		const archiveButton = getByTestId('archive-btn');
		await act(async () => {
			await fireEvent.click(archiveButton);
		});

		expect(archiveMock).toHaveBeenCalledTimes(1);
		expect(notificationSpy).toHaveBeenCalledTimes(1);
		expect(notificationSpy).toHaveBeenCalledWith({
			animationIn: 'animate__bounceIn',
			animationOut: 'animate__bounceOut',
			message: 'Failed to archive mock',
			title: 'Failed to archive mock',
			type: 'danger',
		});
	});

	it('should unarchive mock', async () => {
		const notificationSpy = jest.spyOn(actualNotification, 'default');
		getMockById.mockResolvedValue({
			data: {
				data: {
					...data,
					entityStatus: {
						status: 'ARCHIVED',
					},
				},
			},
		});

		let tree;
		await act(async () => {
			tree = await render(<Detail />);
		});

		const { getByTestId } = tree;

		const unArchiveButton = getByTestId('unarchive-btn');
		await act(async () => {
			await fireEvent.click(unArchiveButton);
		});

		expect(unarchiveMock).toHaveBeenCalledTimes(1);
		expect(notificationSpy).toHaveBeenCalledTimes(1);
		expect(notificationSpy).toHaveBeenCalledWith({
			animationIn: 'animate__bounceIn',
			animationOut: 'animate__bounceOut',
			message: 'Test mock: Mock unarchived successfully',
			title: 'Unarchive success',
			type: 'success',
		});
	});

	it('should show error notification when un-archiving mock fails', async () => {
		const notificationSpy = jest.spyOn(actualNotification, 'default');
		getMockById.mockResolvedValue({
			data: {
				data: {
					...data,
					entityStatus: {
						status: 'ARCHIVED',
					},
				},
			},
		});

		unarchiveMock.mockRejectedValue(new Error('Failed to unarchive mock'));

		let tree;
		await act(async () => {
			tree = await render(<Detail />);
		});

		const { getByTestId } = tree;

		const unArchiveButton = getByTestId('unarchive-btn');
		await act(async () => {
			await fireEvent.click(unArchiveButton);
		});

		expect(unarchiveMock).toHaveBeenCalledTimes(1);
		expect(unarchiveMock).toHaveBeenCalledWith(data.id);

		expect(notificationSpy).toHaveBeenCalledTimes(1);
		expect(notificationSpy).toHaveBeenCalledWith({
			animationIn: 'animate__bounceIn',
			animationOut: 'animate__bounceOut',
			message: 'Failed to unarchive mock',
			title: 'Failed to unarchive mock',
			type: 'danger',
		});
	});

	it('should delete mock', async () => {
		const notificationSpy = jest.spyOn(actualNotification, 'default');
		deleteMockById.mockResolvedValue({});

		let tree;
		await act(async () => {
			tree = await render(<Detail />);
		});

		const { getByTestId } = tree;

		const deleteButton = getByTestId('delete-btn');
		await act(async () => {
			await fireEvent.click(deleteButton);
		});

		const confirmDelete = getByTestId('confirmation-modal-confirm-btn');
		await act(async () => {
			await fireEvent.click(getByTestId('confirmation-modal-cancel-btn'));

			await fireEvent.click(deleteButton);
			await fireEvent.click(confirmDelete);
		});

		expect(deleteMockById).toHaveBeenCalledTimes(1);
		expect(deleteMockById).toHaveBeenCalledWith(data.id);

		expect(notificationSpy).toHaveBeenCalledTimes(1);
		expect(notificationSpy).toHaveBeenCalledWith({
			animationIn: 'animate__bounceIn',
			animationOut: 'animate__bounceOut',
			message: 'Deleted resource successfully!',
			title: 'Deletion success',
			type: 'success',
		});
	});

	it('should show error notification when deleting mock fails', async () => {
		const notificationSpy = jest.spyOn(actualNotification, 'default');
		deleteMockById.mockRejectedValue(new Error('Failed to delete mock'));

		let tree;
		await act(async () => {
			tree = await render(<Detail />);
		});

		const { getByTestId } = tree;

		const deleteButton = getByTestId('delete-btn');
		await act(async () => {
			await fireEvent.click(deleteButton);
		});

		const confirmDelete = getByTestId('confirmation-modal-confirm-btn');
		await act(async () => {
			await fireEvent.click(confirmDelete);
		});

		expect(deleteMockById).toHaveBeenCalledTimes(1);
		expect(deleteMockById).toHaveBeenCalledWith(data.id);

		expect(notificationSpy).toHaveBeenCalledTimes(1);
		expect(notificationSpy).toHaveBeenCalledWith({
			animationIn: 'animate__bounceIn',
			animationOut: 'animate__bounceOut',
			message: 'Failed to delete mock',
			title: 'Failed to delete mock',
			type: 'danger',
		});
	});

	it('should navigate to edit mock page', async () => {
		const mockNavigate = jest.fn();
		useNavigate.mockImplementation(() => mockNavigate);

		let tree;
		await act(async () => {
			tree = await render(<Detail />);
		});

		const { getByTestId } = tree;

		const editButton = getByTestId('edit-btn');
		await act(async () => {
			await fireEvent.click(editButton);
		});

		expect(mockNavigate).toHaveBeenCalledTimes(1);
		expect(mockNavigate).toHaveBeenCalledWith(
			'/mimock-ui/mocks/manage/edit-mock?mockId=' + data.id
		);
	});
});
