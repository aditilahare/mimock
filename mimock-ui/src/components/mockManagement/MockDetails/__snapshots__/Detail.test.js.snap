// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Detail should render detail component 1`] = `
<div>
  <div
    class="flex flex-col justify-start align-middle w-full px-9 mt-4 mb-2"
    data-testid="detail-section"
  />
</div>
`;
